lib/libc/common/source/stdio/feof.c

```diff
@@ -0,0 +1,14 @@
+/*
+ * Copyright (c) 2024 Daniel Hajjar
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdio.h>
+
+int zvfs_feof(FILE *);
+
+int feof(FILE *file)
+{
+	return zvfs_feof(file);
```

Comment:
We shouldn't replace libc's feof here. That was my main concern with this patch. I'm not sure why zvfs is even pretending to offer the FILE type...

Reply:
Should I just remove this from the common libc?

Reply:
> We shouldn't replace libc's feof here. That was my main concern with this patch. I'm not sure why zvfs is even pretending to offer the FILE type...

How would Zephyr manage and associate integer file descriptors to `FILE` objects and vice versa, in that case, @keith-packard?

Reply:
FILE objects are owned by stdio; they contain a bunch of stdio-specific state so that you can call things like getc, fprintf, fread and others. The zvfs FILE objects have no relation to any stdio implementation; I can't see how they are supposed to be used by a standard C application.

Reply:
> How would Zephyr manage and associate integer file descriptors to FILE objects and vice versa, in that case?

This question is still unanswered.

I guess I could ask it differently. What interface would Picolibc prefer for the OS to perform stdio operations on FILE objects? Associate integer file descriptors with FILE objects? It can't be POSIX.

Reply:
The OS cannot perform operations on FILE objects; FILE objects are the purview of the stdio implementation provided by the C library. If you want to perform stdio operations on `fs_file_t` objects, you need to make the stdio implementation inside the C library talk to the Zephyr FS layer. Right now, the only way to do that is to provide POSIX API support from Zephyr so that newlib or picolibc can use their existing POSIX paths.

Picolibc abstracts the OS interfaces so that we could provide stdio on top of `fs_file_t` objects without going through the POSIX layer. For that, we'd create a function in the Picolibc Zephyr support code which took a `fs_file_t` and created a FILE object wrapper. That FILE object would use only Zephyr native interfaces for all operations. Then we could provide an `fopen` implementation also using only Zephyr native interfaces. At that point, we'd have a C compatible stdio implementation with file system access which did not require the POSIX support layer in Zephyr.

I'm happy to write that code; it's not a huge amount of source, but it does require some knowledge of picolibc internals. Is this required to support `fileno`? This would require using fdtable instead of allowing applications to simply allocate `fd_file_t` dynamically as a consequence of calling `fopen`. Right now, fileno doesn't work on stdin/stdout/stderr, and only works when applications use fopen and the POSIX Zephyr support.

Reply:
It seems like what's being proposed is an interface which isn't dependent on POSIX would be exposed for picolibc instead of reimplementing functions like `feof`, if I'm understanding this correct.

If that's the case, then this should just be closed, right?